examples/vision/super_resolution_sub_pixel.py

```diff
@@ -0,0 +1,513 @@
+"""
+Title: Super-Resolution Using an Efficient Sub-Pixel CNN
+Author: [Xingyu Long](https://github.com/xingyu-long)
+Date created: 2020/07/28
+Last modified: 2020/07/30
+Description: Implementing Super-Resoluion using Efficient sub-pixel model on BSDS500.
+"""
+
+"""
+## Introduction
+
+ESPCN (Efficient Sub-Pixel CNN), proposed by [Shi, W.
+2016](https://arxiv.org/abs/1609.05158) is used to reconstruct the image from low
+resolution (LR) to high resolution (HR). In this paper, they introduce an efficient
+sub-pixel convolution layers which learns an array of upscaluing filter to upscale the
+final LR feature maps into the HR output. This post will implement the model in this
+paper and trained with relatively small dataset
+[BSDS500](http://www.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/BSR/BSR_bsds500.tgz).
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+
+"""
+## Load data: BSDS500 dataset
+
+### Download data
+
+We use Keras built-in util `keras.utils.get_file` to get the dataset.
+"""
+
+dataset_url = "http://www.eecs.berkeley.edu/Research/Projects/CS/vision/grouping/BSR/BSR_bsds500.tgz"
+data_dir = tf.keras.utils.get_file(origin=dataset_url, fname="BSR", untar=True)
+
+"""
+Check the `data_dir` and build data path.
+"""
+
+import os
+
+BSDS500 = os.path.join(data_dir, "BSDS500/data")
+print(BSDS500)
+
+"""
+Prepare train and test dataset path and check the number of train, test and validation
+samples.
+"""
+
+dataset = os.path.join(BSDS500, "images")
+train_path = os.path.join(dataset, "train")
+test_path = os.path.join(dataset, "test")
+valid_path = os.path.join(dataset, "val")
+
+input_img_paths = sorted(
+    [
+        os.path.join(train_path, fname)
+        for fname in os.listdir(train_path)
+        if fname.endswith(".jpg")
+    ]
+)
+
+test_img_paths = sorted(
+    [
+        os.path.join(test_path, fname)
+        for fname in os.listdir(test_path)
+        if fname.endswith(".jpg")
+    ]
+)
+
+valid_img_paths = sorted(
+    [
+        os.path.join(valid_path, fname)
+        for fname in os.listdir(valid_path)
+        if fname.endswith(".jpg")
+    ]
+)
+
+print("Number of train samples: ", len(input_img_paths))
+print("Number of test samples: ", len(test_img_paths))
+print("Number of validation samples: ", len(valid_img_paths))
+
+"""
+## Visualiza the data
+"""
+
+from IPython.display import display
+from tensorflow.keras.preprocessing.image import load_img
+
+display(load_img(input_img_paths[0]))
+
+"""
+## Crop and resize images.
+
+Since we need to upscale images and we have to process the input data format. First, you
+need to define crop size and calculate the acutal crop size by following functions. For
+input data, after we get the crop image and blur it with `bicubic` method. For target
+data, we only do the crop operation.
+"""
+
+import PIL
```

Comment:
Rather than doing everything in PIL, which won't scale very well and takes up a lot of code, prefer creating a tf.data.Dataset (e.g. with `image_dataset_from_directory()` and process images with TF ops.

You can use https://www.tensorflow.org/api_docs/python/tf/image/rgb_to_yuv and its inverse yuv_to_rgb

Probably something like:

```python
def process_input(x):
   return tf.image.resize(x, ...)

def process_target(x):
   return tf.image.crop(x, ...)

dataset = image_dataset_from_directory(directory, label_mode=None)
dataset = dataset.map(rgb_to_yuv)
dataset = dataset.map(lambda x: (process_input(x), process_target(x)))
```

Reply:
+1 on this! `tf.data.Dataset` is the recommended API for the input pipeline of TF 2.x. This will be a good showcase of it.